### ISSUE TITLE
Follow semantic releases (semver) specification

### ISSUE THREAD

**Issue by darkowic** [okta/okta-react#35]
https://semver.org/

Why do you introduce breaking changes in the bugfix version? In 3.0.4 one of the deprecated API has been removed **without even marking this as breaking change - just a 'bugfix'** - https://github.com/okta/okta-react/releases/tag/%40okta%2Fokta-react-3.0.4. Also, it still exists in README which is misleading. 

It will really help everyone avoid bugs and a bad user experience.

**Comment by aarongranick-okta** [comment]
@darkowic Thank you for reaching out. We do take semantic versioning seriously and strive to always follow the guidelines. In this case we removed the **default** implementation of `onSessionExpired`. We had quite  bit of internal debate about this, but the bottom line is that this method was never doing what we intended it to do and was causing many other issues (concurrency / race conditions). Because this version of the SDK uses a "passive" auto renew, the method would only be called when a token is being read. This is typically when navigating to a secure route. Because the default logic would trigger a redirect flow, it would cause the context of the user navigation to be lost and the user would end up at the app's home page.  Additionally, it caused a race condition because it was starting an authorization flow, while in the process of evaluating whether an authorization flow should be started.  Because the default implementation caused so many issues, we considered removing it to be a bug fix.

In this version of the SDK, you **can** provide your own implementation of `onSessionExpired` but we do not recommend it. In the next version of the SDK this option has been removed completely. The `onAuthRequired` or `isAuthenticated` options should be able to provide the same level of control without creating race conditions.

What were you relying on `onSessionExpired` to provide? There is probably a better way to get it. I'd be happy to help point you in that direction.

**Comment by darkowic** [comment]
We are maintaining an internal framework and we do not introduce breaking changes in minor releases. TBH I didn't expect that anything will be just removed (especially if it is still in docs(!)) even if doesn't work as expected. It may be breaking for someone. It blocks us now to upgrade the package to a newer version where other bugs are fixed because we have to fix this first.

We are hiding the okta provider under our own provider. As I understand, we have to now implement our own `onAuthSessionExpired` if we still want to support it.

**Comment by aarongranick-okta** [comment]
@darkowic The **option** for `onSessionExpired` is not removed. If any user has passed this option, the function will still be called when an access token renewal fails. What was removed was a **default handler** which would begin a redirect login flow (without any context). This almost always occurred while evaluating `isAuthenticated` by a `SecureRoute`, whose next step would be to start a login flow (with route context). So the effect of the default handler was to change/break the behavior of the `SecureRoute` when the access token is expired OR if 3rd party cookies are blocked (which always causes token renew failure, even though the session is valid). 

A large part of user confusion was based around the idea that this method actually does what it is named - that it provides a callback when the user session is expired. THIS IS FALSE. The callback is called when an access token renewal fails and this only occurs when a token is being READ. It will not occur in the background or any other time outside a token renew, which (with `autoRenew` option, true by default) usually occurs while calling `tokenManager.get('accessToken')`. With this handler in place, sometimes reading a token will return a value and sometimes it will redirect to Okta (!!) - This would break logic in `isAuthenticated` or any other code which makes decisions based upon the presence of the access token. As browser vendors began to block 3rd party cookies by default this issue became more prevalent and we made the decision to remove the default handler. This fixes the behavior of `tokenManager.get()` so it will either return a value or undefined, as it had before the poorly-named and conceived `onSessionExpired` option was added. This also fixes the issues with `isAuthenticated` and `SecureRoute`

If you feel like your library should preserve this behavior, the previous code is here: https://github.com/okta/okta-oidc-js/pull/848/files#diff-6ed2d59603d0bcdaee5fd7ed53ff86f262a7be217924e20ab9df2cef4fe433dcL35

If no `onSessionExpired` option was passed, it set the `login` function as the option.

The README of `okta-auth-js` (on which okta-react is based) has a little more information: https://github.com/okta/okta-auth-js/tree/3.2#additional-options  If you'd like to track token renew failures, there is an event interface on the `TokenManager`. Unless you are implementing your own renew logic, we don't recommend taking any direct action on token renew failures. In this version of the SDK, the event will only fire during renew, which only occurs while reading a token from the `TokenManager`. Performing any logic on this event may interrupt the caller from being able to read the token value. 

Normally `isAuthenticated` will be called by `SecureRoute`, and seeing a false value it will begin a new authentication flow with the proper context to return to the correct route.
